crawler/src/main.rs

```diff
@@ -0,0 +1,299 @@
+                node.send_broadcast(ClientMessage::PeerRequest);
+                tokio::time::sleep(Duration::from_secs(PEER_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Starts the usual periodic activities of a crawler node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+    }
+}
+
+/// Inbound message processing logic for the test nodes.
+#[async_trait::async_trait]
+impl Reading for Crawler {
+    type Message = ClientMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        // FIXME: use the maximum message size allowed by the protocol or (better) use streaming deserialization.
+        let mut buf = [0u8; 8 * 1024];
+
+        reader.read_exact(&mut buf[..MESSAGE_LENGTH_PREFIX_SIZE])?;
+        let len = u32::from_le_bytes(buf[..MESSAGE_LENGTH_PREFIX_SIZE].try_into().unwrap()) as usize;
+
+        if len > buf.len() {
+            return Ok(None);
+        }
+
+        if reader.read_exact(&mut buf[..len]).is_err() {
+            return Ok(None);
+        }
+
+        match ClientMessage::deserialize(&buf[..len]) {
+            Ok(msg) => {
+                info!(parent: self.node().span(), "received a {} from {}", msg.name(), source);
+                Ok(Some(msg))
+            }
+            Err(e) => {
+                error!("a message from {} failed to deserialize: {}", source, e);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        }
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        // We may need to handle other messages as well.
+        match message {
+            ClientMessage::BlockRequest(_start_block_height, _end_block_height) => {}
+            ClientMessage::BlockResponse(_block) => {}
+            ClientMessage::Disconnect => {}
+            ClientMessage::PeerRequest => self.process_peer_request(source).await?,
+            ClientMessage::PeerResponse(peer_ips) => self.process_peer_response(source, peer_ips).await?,
+            ClientMessage::Ping(version, _fork_depth, _peer_type, _peer_state, _block_hash, block_header) => {
+                // Deserialise the block header.
+                let block_header = block_header.deserialize().await.unwrap();
+                self.process_ping(source, version, block_header.height()).await?
+            }
+            ClientMessage::Pong(_is_fork, _block_locators) => {}
+            ClientMessage::UnconfirmedBlock(_block_height, _block_hash, _block) => {}
+            ClientMessage::UnconfirmedTransaction(_transaction) => {}
+            _ => return Err(io::ErrorKind::InvalidData.into()), // Peer is not following the protocol.
+        }
+
+        Ok(())
+    }
+}
+
+impl Crawler {
+    async fn process_peer_request(&self, source: SocketAddr) -> io::Result<()> {
+        // Don't share peers for the time being.
```

Comment:
needs a `TODO` for easier tracking

Reply:
We may just want to delete this—there's no reason for the crawler to share peers (unless we hope that this could increase the connectivity of the network) 🤔